src/plots/cartesian/autorange.js

```diff
@@ -0,0 +1,324 @@
+        if(ax.type === 'linear' || ax.type === '-') {
+            if(ax.rangemode === 'tozero') {
+                if(minbest.val >= 0) {
+                    minbest = {val: 0, pad: 0};
+                }
+                if(maxbest.val <= 0) {
+                    maxbest = {val: 0, pad: 0};
+                }
+            }
+            else if(ax.rangemode === 'nonnegative') {
+                if(minbest.val - mbest * getPad(minbest) < 0) {
+                    minbest = {val: 0, pad: 0};
+                }
+                if(maxbest.val < 0) {
+                    maxbest = {val: 1, pad: 0};
+                }
+            }
+
+            // in case it changed again...
+            mbest = (maxbest.val - minbest.val) /
+                (ax._length - getPad(minbest) - getPad(maxbest));
+
+        }
+
+        newRange = [
+            minbest.val - mbest * getPad(minbest),
+            maxbest.val + mbest * getPad(maxbest)
+        ];
+    }
+
+    // don't let axis have zero size, while still respecting tozero and nonnegative
+    if(newRange[0] === newRange[1]) {
+        if(ax.rangemode === 'tozero') {
+            if(newRange[0] < 0) {
+                newRange = [newRange[0], 0];
+            }
+            else if(newRange[0] > 0) {
+                newRange = [0, newRange[0]];
+            }
+            else {
+                newRange = [0, 1];
+            }
+        }
+        else {
+            newRange = [newRange[0] - 1, newRange[0] + 1];
+            if(ax.rangemode === 'nonnegative') {
+                newRange[0] = Math.max(0, newRange[0]);
+            }
+        }
+    }
+
+    // maintain reversal
+    if(axReverse) newRange.reverse();
+
+    return Lib.simpleMap(newRange, ax.l2r || Number);
+}
+
+/*
+ * calculate the pixel padding for ax._min and ax._max entries with
+ * optional extrapad as 5% of the total axis length
+ */
+function makePadFn(ax) {
+    // 5% padding for points that specify extrapad: true
+    var extrappad = ax._length / 20;
+
+    // domain-constrained axes: base extrappad on the unconstrained
+    // domain so it's consistent as the domain changes
+    if((ax.constrain === 'domain') && ax._inputDomain) {
+        extrappad *= (ax._inputDomain[1] - ax._inputDomain[0]) /
+            (ax.domain[1] - ax.domain[0]);
+    }
+
+    return function getPad(pt) { return pt.pad + (pt.extrapad ? extrappad : 0); };
+}
+
+function doAutoRange(ax) {
+    ax.setScale();
+
+    // TODO do we really need this?
+    var hasDeps = (ax._min && ax._max && ax._min.length && ax._max.length);
+
+    if(ax.autorange && hasDeps) {
+        ax.range = getAutoRange(ax);
+
+        ax._r = ax.range.slice();
+        ax._rl = Lib.simpleMap(ax._r, ax.r2l);
+
+        // doAutoRange will get called on fullLayout,
+        // but we want to report its results back to layout
+
+        var axIn = ax._input;
+        axIn.range = ax.range.slice();
+        axIn.autorange = ax.autorange;
+    }
+}
+
+function needsAutorange(ax) {
```

Comment:
It feels awkward to export this now that auto-range routines are out of `axes.js`.

Maybe we should move [`fastAxisExpand`](https://github.com/plotly/plotly.js/blob/8731cb5d87f305fff8386b5d8eb2f0280752393f/src/traces/scattergl/index.js#L144-L161) out of scattergl/index.js and into this file?

Reply:
Or just bake the `fastAxisExpand` functionality into `expand` to be used automatically when the paddings are constant - then everyone wins! I think that should be possible... I'll try it.

Reply:
> then everyone wins!

That's an idea! But perhaps the [`TOO_MANY_POINTS`](https://github.com/plotly/plotly.js/blob/8731cb5d87f305fff8386b5d8eb2f0280752393f/src/traces/scattergl/index.js#L42) constant should remain per trace type?